group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java

```diff
@@ -0,0 +1,261 @@
+                // Only topics that are present in both the subscribed topics list and the topic metadata should be considered for assignment.
+                if (assignmentSpec.topics().containsKey(topicId)) {
+                    membersPerTopic
+                        .computeIfAbsent(topicId, k -> new ArrayList<>())
+                        .add(memberId);
+                } else {
+                    log.warn(memberId + " subscribed to topic " + topicId + " which doesn't exist in the topic metadata");
+                }
+            }
+        });
+
+        return membersPerTopic;
+    }
+
+    private Map<Uuid, List<Integer>> getUnassignedPartitionsPerTopic(final AssignmentSpec assignmentSpec, Map<Uuid, Set<Integer>> assignedStickyPartitionsPerTopic) {
+        Map<Uuid, List<Integer>> unassignedPartitionsPerTopic = new HashMap<>();
+        Map<Uuid, AssignmentTopicMetadata> topicsMetadata = assignmentSpec.topics();
+
+        topicsMetadata.forEach((topicId, assignmentTopicMetadata) -> {
+            ArrayList<Integer> unassignedPartitionsForTopic = new ArrayList<>();
+            int numPartitions = assignmentTopicMetadata.numPartitions();
+            // List of unassigned partitions per topic contains the partitions in ascending order.
+            Set<Integer> assignedStickyPartitionsForTopic = assignedStickyPartitionsPerTopic.getOrDefault(topicId, new HashSet<>());
+            for (int i = 0; i < numPartitions; i++) {
+                if (!assignedStickyPartitionsForTopic.contains(i)) {
+                    unassignedPartitionsForTopic.add(i);
+                }
+            }
+            unassignedPartitionsPerTopic.put(topicId, unassignedPartitionsForTopic);
+        });
+
+        return unassignedPartitionsPerTopic;
+    }
+
+/**
+ * <p> The algorithm includes the following steps:
+ * <ol>
+ * <li> Generate a map of <code>membersPerTopic</code> using the given member subscriptions.</li>
+ * <li> Generate a list of members (<code>potentiallyUnfilledMembers</code>) that have not met the minimum required quota for assignment AND
+ * get a list (<code>assignedStickyPartitionsPerTopic</code>) of partitions that we want to retain in the new assignment.</li>
+ * <li> Add members from the <code>potentiallyUnfilledMembers</code> list to the <code>unfilledMembers</code> list if they haven't met the total required quota i.e. minimum number of partitions per member + 1 (if member is designated to receive one of the excess partitions). </li>
+ * <li> Generate a list of unassigned partitions by calculating the difference between total partitions and already assigned (sticky) partitions. </li>
+ * <li> Iterate through the unfilled members and assign partitions from the unassigned partitions. </li>
+ * </ol>
+ * </p>
+ */
+    @Override
+    public GroupAssignment assign(final AssignmentSpec assignmentSpec) throws PartitionAssignorException {
+        Map<String, Map<Uuid, Set<Integer>>> membersWithNewAssignmentPerTopic = new HashMap<>();
```

Comment:
Would it be possible to directly use `MemberAssignment` instead of `Map<Uuid, Set<Integer>` here? That would save allocating a HashMap at the end.

Reply:
We need to able to modify the Map<Uuid, Set<Integer>> throughout the code as we assign partitions and since targetPartitions in MemberAssignment is private final we can't modify it once its initialized. This is why I had to do it this way.

Reply:
This is not entirely correct. Yes, the assignment is private final but this does not prevent you from mutating the hash map. It only prevents you from re-assigning the attribute.

Reply:
oh sorry my bad, I've always learnt that final means you can't modify the value after, but I guess for a map you can't modify the reference but can change the values. I'll change it thanks!